Add CI jobs to run in non UTC timezone

notatallshaw opened the pull request:
Adds tests for #10816 and requires #10834 to pass

I originally tried to add "timezone" to the existing matrix for Windows tests but given it only runs for Python 3.10 Unit Tests it created a bunch of tasks in the github workflow that didn't do anything except create a machine, set up the environment, and then shutdown.

notatallshaw commented:
Get the following error trying to run:

> szenius/set-timezone@v1.0 is not allowed to be used in pypa/pip. Actions in this workflow must be: within a repository owned by pypa, created by GitHub or match the following: actions/checkout@v2, actions/setup-python@v2, dessant/lock-threads@v3, dorny/paths-filter@v2, pradyunsg/auto-label-merge-conflicts@v3, pre-commit/action@v2.0.0.

But I have tested this does produced the expected failures on my own repository: https://github.com/notatallshaw/pip/runs/4972090301?check_suite_focus=true

pradyunsg commented on .github/workflows/ci.yml:
```diff
@@ -218,3 +218,60 @@ jobs:
+      - uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python }}
+      - uses: szenius/set-timezone@v1.0
```
```suggestion
      - uses: pradyunsg/set-timezone@v1.0
```

I've forked the package and allowed that fork, to reduce the potential number of users that could hijack what code is run in our CI pipeline.

The current list of allowed actions is:

actions/checkout@v2,
actions/setup-python@v2,
dessant/lock-threads@v3,
dorny/paths-filter@v2,
pradyunsg/auto-label-merge-conflicts@v3,
pradyunsg/set-timezone@v1.0,
pre-commit/action@v2.0.0,

Both paths-filter and lock-threads have decently large userbases, so I'm fine with the risk we take on via those.

notatallshaw commented:
Unit tests failed as expected: https://github.com/pypa/pip/runs/4981934385?check_suite_focus=true

These are the same failing unit tests I see when I run pips unit tests on my local machine.

notatallshaw commented:
Not sure if it's overkill but I've switched from just testing Singapore timezone to testing US Pacific and New Zealand time zones,

I picked these two because if someone picks a UTC datetime one of these timezones is likely going to be on a different date, so even if the formatting of the datetime only includes date part and not the time part this will likely catch that. I could of picked more extreme time zones but in my experience working with time zones more extreme than these two is that the relevant country can change what side of the international dateline they are on.

I also added a link to the Microsoft Docs on time zones because Microsoft uses non-standard time zones names.

Happy to revert either of these changes if you think they are too much.

notatallshaw commented:
Also learning a bit about GitHub Actions I can use Includes and/or Excludes instead of copying the whole Windows set-up, but it does add quite a bit of complexity to the CI matrix logic: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-including-additional-values-in-combinations
